File: iwapp_sandp/public/js/purchase_receipt.js
frappe.ui.form.on('Purchase Receipt', {
    custom_site_address: function (frm) {
        dislpay_site_address(frm)
    },
    custom_customer: function (frm) {
        frm.set_value("custom_site_address", "")
        set_site_address_filter(frm)
    },
    refresh: function (frm) {
        dislpay_site_address(frm)
        set_site_address_filter(frm)
        if (frm.doc.docstatus != 1) {
            frm.add_custom_button(__('Create Model Id'), function () {
                if (frm.doc.items) {
                    frm.clear_table("custom_purchase_item_entry")
                    frm.refresh_field("custom_purchase_item_entry")
                    $.each(frm.doc.items, function (idx, item) {
                        if (item.custom_has_model_id == 1) {
                            var child = frm.add_child("custom_purchase_item_entry");
                            child.purchase_item = item.item_code
                            child.item_qty = item.qty
                            child.model_number = item.custom_model_id
                            // child.item_rate = item.rate
                            // child.discount_percentage = item.discount_percentage
                            child.unit = item.uom
                            child.has_serial_no = item.custom_has_serial_no
                            frm.refresh_fields("custom_purchase_item_entry");
                        }
                    })
                }
            })
        }
        frm.fields_dict['custom_purchase_item_entry'].grid.get_field('model_number').get_query = function (doc, cdt, cdn) {
            var child = locals[cdt][cdn];
            return {
                filters: {
                    'item': child.purchase_item
                }
            };
        };
        frm.fields_dict['items'].grid.get_field('custom_model_id').get_query = function (doc, cdt, cdn) {
            var child = locals[cdt][cdn];
            return {
                filters: {
                    'item': child.item_code
                }
            };
        };
    }
});

frappe.ui.form.on('Purchase Receipt Item', {
    custom_model_id: function (frm, cdt, cdn) {
        var item = locals[cdt][cdn];
        frappe.model.set_value(cdt, cdn, "description", "")
        if (item.custom_model_id) {
            frappe.db.get_value('Item Model ID', item.custom_model_id, ["brand", "specification"]).then(response => {
                const brand_and_spec = response.message;
                if (brand_and_spec) {
                    const brand = brand_and_spec.brand || "";
                    const specification = brand_and_spec.specification || "";
                    if (specification) {
                        const description = `${item.item_name}, ${brand} - ${item.custom_model_id}, ${specification}`;
                        frappe.model.set_value(cdt, cdn, "description", description)
                    } else {
                        const description = `${item.item_name}, ${brand} - ${item.custom_model_id}`;
                        frappe.model.set_value(cdt, cdn, "description", description)
                    }
                }
            });
        }
    },
    item_code: function (frm, cdt, cdn) {
        var item = locals[cdt][cdn];
        if (item.custom_model_id) {
            frappe.model.set_value(cdt, cdn, "custom_model_id", "")
        }
    }
    // custom_model_id: function (frm, cdt, cdn) {
    // let row = locals[cdt][cdn];
    // if (row.custom_model_id) {
    //     frappe.db.get_value('Item Model ID', row.custom_model_id, ['brand', 'specification'])
    //         .then(r => {
    //             let values = r.message;
    //             var result = row.item_name + "-" + values.specification + "-" + values.brand + "-" + row.custom_model_id;
    //             frappe.model.set_value(row.doctype, row.name, "description", result)
    //         })
    // }
    // }
})

var set_site_address_filter = function (frm) {
    if (frm.doc.custom_customer) {
        frm.set_query('custom_site_address', function (doc) {
            return {
                filters: {
                    'link_doctype': 'Customer',
                    'link_name': doc.custom_customer
                }
            }
        })
    }
}

var dislpay_site_address = function (frm) {
    if (frm.doc.custom_site_address) {
        // Fetch the Address document based on the custom_site_address field
        frappe.db.get_doc('Address', frm.doc.custom_site_address)
            .then(doc => {
                // Initialize an array to hold the address components
                let address_components = [doc.address_line1];

                // Always add city
                address_components.push(doc.city);

                // Check if county and city are the same, ignoring case sensitivity
                if (doc.county && doc.city && doc.county.toLowerCase() !== doc.city.toLowerCase()) {
                    address_components.push(doc.county);
                }

                if (doc.state) {
                    address_components.push(doc.state);
                }

                if (doc.country) {
                    address_components.push(doc.country);
                }

                // Add doc.pincode to the address components if it exists, formatted with a dash and a space
                if (doc.pincode) {
                    address_components.push(`- ${doc.pincode}`);
                }

                // Join the address components with a comma and space, but leave out the final comma
                const site_address_display = address_components.join(', ').replace(/, (?=\-)/, ' ');

                // Set the combined string to the HTML field
                frm.set_df_property('custom_site_address_html', 'options', site_address_display);
            });
    }
}

File: iwapp_sandp/events/purchase_receipt.py
import frappe
from bs4 import BeautifulSoup

def validate(doc, method):
    if doc.custom_purchase_item_entry:
        for model in doc.custom_purchase_item_entry:
            for item in doc.items:
                if item.item_code ==  model.purchase_item:
                    item.custom_model_id = model.model_number
                    item.qty = model.item_qty
                    # item.rate = model.item_rate
                    # item.discount_percentage = model.discount_percentage
                    item.uom = model.unit
                    item.serial_no = model.serial_number

def before_save(doc, method):
    if doc.items:
        for item in doc.items:
            if item.custom_model_id and item.uom != "Nos":
                # To set custom_has_model_id = 1 when creating model id from PR
                item.custom_has_model_id = 1
                item.custom_has_batch_no = 1
                frappe.db.set_value("Item", item.item_code, {"custom_has_model_id" : 1, "has_batch_no" : 1}, update_modified = False)
                # get_list items having BATCH No SERIES BLANK
                model_id_items = frappe.db.get_value("Item", {
                        "name":item.item_code,
                        "batch_number_series": ""}, "name"
                )
                if model_id_items:
                    # Take the first 3 letters from itemcode and with yy-mm-dd and sb set to serial_no_series in ITEM
                    first_three_letters = item.item_code[:3].upper()
                    batch_no_series = f"{first_three_letters}.YY.MM.DD.####"
                    frappe.db.set_value("Item", item.item_code, {"custom_has_model_id" : 1, "has_batch_no" : 1, "create_new_batch" :1,
                    "batch_number_series" : batch_no_series})
            if item.custom_model_id and item.uom == "Nos":
                 # To set custom_has_model_id = 1 when creating model id from PR
                item.custom_has_model_id = 1
                item.custom_has_serial_no = 1
                frappe.db.set_value("Item", item.item_code, {"custom_has_model_id" : 1, "has_serial_no" : 1}, update_modified = False)
                # get_list items having BATCH No SERIES BLANK
                model_id_items = frappe.db.get_value("Item",{
                        "name":item.item_code,
                        "serial_no_series": ""},"name")
                if model_id_items:
                    # Take the first 3 letters from itemcode and with yy-mm-dd and sb set to serial_no_series in ITEM
                    first_three_letters = item.item_code[:3].upper()
                    serial_no_series = f"{first_three_letters}.YY.MM.DD.####"
                    frappe.db.set_value("Item", item.item_code, {"custom_has_model_id" : 1, "has_serial_no" : 1,
                    "serial_no_series" : serial_no_series})
            if item.custom_from_model_id == 0:
                if item.custom_model_id:
                    brand_and_spec = frappe.db.get_value('Item Model ID', item.custom_model_id, ["brand", "specification"])
                    if brand_and_spec:
                        brand = brand_and_spec[0] if brand_and_spec[0] else ""
                        specification = brand_and_spec[1] if brand_and_spec[1] else ""
                        if specification:
                            item.description = f"{item.item_name}, {brand} - {item.custom_model_id}, {specification}"
                        else:
                            item.description = f"{item.item_name}, {brand} - {item.custom_model_id}"
                        item.brand = brand
                        item.custom_from_model_id = 1

def on_submit(doc, method):
    if doc.items:
        for item in doc.items:
            if item.custom_has_serial_no == 1 and item.serial_no:
                frappe.db.set_value("Item Model ID", item.custom_model_id, "serial_no", item.serial_no)
                # Replace commas with newline characters and split the string into a list and Remove spaces from each element in the list
                data_list = [i.strip() for i in item.serial_no.replace(',', '\n').split('\n') if i]
                for i in data_list:
                    frappe.db.set_value("Serial No", i, {"custom_model_id" : item.custom_model_id, "description" : item.description,
                        "brand":item.brand, "custom_update_model_id":1})
            if item.custom_has_batch_no == 1 and item.custom_model_id:
                frappe.db.set_value("Batch", item.batch_no, {"custom_model_id" : item.custom_model_id, "description" : item.description, "custom_brand":item.brand})
                # Replace commas with newline characters and split the string into a list and Remove spaces from each element in the list
                # data_list = [i.strip() for i in item.serial_no.replace(',', '\n').split('\n') if i]
                # for i in data_list:
                    # frappe.db.set_value("Serial No", i, {"custom_model_id" : item.custom_model_id, "description" : item.description, "brand":item.brand, "custom_update_model_id":1})

